sha2/src/sha512/succinct.rs

```diff
@@ -126,43 +97,37 @@ fn sha512_compress_stub(w: &mut [u64; 80], state: &mut [u64; 8]) {
         0x6c44198c4a475817,
     ];
 
-    let mut a = state[0];
-    let mut b = state[1];
-    let mut c = state[2];
-    let mut d = state[3];
-    let mut e = state[4];
-    let mut f = state[5];
-    let mut g = state[6];
-    let mut h = state[7];
+    for i in 0..blocks.len() {
+        let mut w = [0u64; 80];
+        for j in 0..16 {
+            w[j] = u64::from_be_bytes([
+                blocks[i][j * 8 + 0],
+                blocks[i][j * 8 + 1],
+                blocks[i][j * 8 + 2],
+                blocks[i][j * 8 + 3],
+                blocks[i][j * 8 + 4],
+                blocks[i][j * 8 + 5],
+                blocks[i][j * 8 + 6],
+                blocks[i][j * 8 + 7],
+            ]);
+        }
+
+        for i in 16u32..80u32 {
+            unsafe {
+                syscall_sha512_extend(w.as_mut_ptr(), i);
+            }
+        }
 
-    for i in 0..80 {
-        let s1 = e.rotate_right(14) ^ e.rotate_right(18) ^ e.rotate_right(41);
-        let ch = (e & f) ^ (!e & g);
-        let temp1 = h
-            .wrapping_add(s1)
-            .wrapping_add(ch)
-            .wrapping_add(k[i])
-            .wrapping_add(w[i]);
-        let s0 = a.rotate_right(28) ^ a.rotate_right(34) ^ a.rotate_right(39);
-        let maj = (a & b) ^ (a & c) ^ (b & c);
-        let temp2 = s0.wrapping_add(maj);
+        zkvm_state[8] = 0; // set initial `i`
+        for _ in 0..80 {
+            unsafe {
+                syscall_sha512_compress(w.as_mut_ptr(), zkvm_state.as_mut_ptr());
```

Comment:
We can replace `zkvm_state` with a pointer to `(i, &k[i])`

Reply:
I've opted to instead use a global `ZKVM_STATE` array for now, which enables sharing of the `K` constants between all `compress` calls.